File: backends/dotnet/src/Zappy.Domain/Shared/Money.cs
namespace Zappy.Domain;

public sealed record Money
{
    public const string EuroCurrency = "EUR";

    public Money(int amount, string currency)
    {
        if (amount < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(amount), amount, "An amount of money is never negative.");
        }

        if (string.IsNullOrWhiteSpace(currency))
        {
            throw new ArgumentException("An amount of money always names its currency.", nameof(currency));
        }

        Amount = amount;
        Currency = currency;
    }

    public int Amount { get; }

    public string Currency { get; }

    public bool IsZero => Amount == 0;

    public static Money Euro(int amount) => new(amount, EuroCurrency);

    public static Money ZeroIn(string currency) => new(0, currency);

    public Money Plus(Money other) => new(Amount + AmountOf(other), Currency);

    public Money Minus(Money other) => new(Amount - AmountOf(other), Currency);

    public Money Times(int factor) => new(Amount * factor, Currency);

    public bool IsAtLeast(Money other) => Amount >= AmountOf(other);

    public Money CappedAt(Money ceiling) => IsAtLeast(ceiling) ? ceiling : this;

    public override string ToString() => $"{Amount} {Currency}";

    private int AmountOf(Money other) => other.Currency == Currency
        ? other.Amount
        : throw new InvalidOperationException($"An amount in {Currency} cannot be combined with an amount in {other.Currency}.");
}

File: backends/java/zappy-domain/src/main/java/com/zappymart/domain/shared/Money.java
package com.zappymart.domain.shared;

import java.util.Objects;

public record Money(int amount, String currency) {

    public static final String EURO = "EUR";

    public Money {
        Objects.requireNonNull(currency, "currency");
        if (amount < 0) {
            throw new IllegalArgumentException("An amount of money is never negative: " + amount);
        }
        if (currency.isBlank()) {
            throw new IllegalArgumentException("A currency code is never blank");
        }
    }

    public static Money euro(int amount) {
        return new Money(amount, EURO);
    }

    public static Money zero() {
        return euro(0);
    }

    public Money plus(Money other) {
        requireSameCurrency(other);
        return new Money(amount + other.amount, currency);
    }

    public Money minus(Money other) {
        requireSameCurrency(other);
        return new Money(amount - other.amount, currency);
    }

    public Money times(int factor) {
        if (factor < 0) {
            throw new IllegalArgumentException("An amount of money is never multiplied by a negative factor: " + factor);
        }
        return new Money(amount * factor, currency);
    }

    public Money percentageRoundedHalfUp(int percentage) {
        if (percentage < 0 || percentage > 100) {
            throw new IllegalArgumentException("A percentage lies between 0 and 100: " + percentage);
        }
        return new Money((amount * percentage + 50) / 100, currency);
    }

    public Money cappedAt(Money ceiling) {
        requireSameCurrency(ceiling);
        return amount <= ceiling.amount ? this : ceiling;
    }

    public boolean isAtLeast(Money other) {
        requireSameCurrency(other);
        return amount >= other.amount;
    }

    public boolean isLessThan(Money other) {
        return !isAtLeast(other);
    }

    public boolean isZero() {
        return amount == 0;
    }

    private void requireSameCurrency(Money other) {
        Objects.requireNonNull(other, "other");
        if (!currency.equals(other.currency)) {
            throw new IllegalArgumentException(
                    "Two amounts in different currencies do not meet: " + currency + " and " + other.currency);
        }
    }
}
